src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -86,12 +88,31 @@ public class TopicPartitionChannel {
   // This offset is updated when Snowflake has received offset from insertRows API
   // We will update this value after calling offsetToken API for this channel
   // We will only update it during start of the channel initialization
-  private long offsetPersistedInSnowflake = NO_OFFSET_TOKEN_REGISTERED_IN_SNOWFLAKE;
+  private AtomicLong offsetPersistedInSnowflake =
+      new AtomicLong(NO_OFFSET_TOKEN_REGISTERED_IN_SNOWFLAKE);
 
   // used to communicate to the streaming ingest's insertRows API
   // This is non final because we might decide to get the new instance of Channel
   private SnowflakeStreamingIngestChannel channel;
 
+  /**
+   * Offsets are reset in kafka when one of following cases arises in which we rely on source of
+   * truth (Which is Snowflake's committed offsetToken)
+   *
+   * <p>1. If channel fails to fetch offsetToken in kafka, we reopen the channel and try to fetch
+   * offset from Snowflake again
+   *
+   * <p>2. If channel fails to ingest a buffer(Buffer containing rows/offsets), we reopen the
+   * channel and try to fetch offset from Snowflake again
+   *
+   * <p>In both above cases, we ask Kafka to send back offsets after offset already present in
```

Comment:
Nit, I think you're missing some words after "send back offsets"

Reply:
No, I think it is meant to be like that. Let me make it more clear. 
`In both above cases, we ask Kafka to send back offsets, strictly from offset number after the offsetToken present in Snowflake`

I will replace it.